### PR TITLE
Drivertrain

### PR DESCRIPTION
**Pull request by billwpierce**
Added code that will make the drivertrain run using the driving controls for offseason-2015A

TODO: Correct ports to those on the 2017 robot.

### CODE REVIEW

File: src/org/usfirst/frc4904/robot/RobotMap.java [hunk 1]
```diff
@@ -57,55 +68,89 @@
...
-		Component.ballIntake = new BallInnie(new VictorSP(Port.PWM.ballInnie));
-		// Ball Dumper
-		Component.ballDumper = new Dump(new VictorSP(Port.PWM.vomitElevator), new VictorSP(Port.PWM.vomitOuttakeRoller));
-		// Flywheel
-		Motor leftFlywheelMotor = new Motor(new VictorSP(Port.PWM.flywheelLeftMotor));
-		leftFlywheelMotor.setInverted(true);
-		Motor rightFlywheelMotor = new Motor(new VictorSP(Port.PWM.flywheelRightMotor));
-		rightFlywheelMotor.setInverted(false);
-		Component.flywheel = new Flywheel(leftFlywheelMotor, rightFlywheelMotor);
+		double inchesPerPulse = 2 * Math.PI * Constant.Chassis.WHEEL_RADIUS_INCHES / Constant.Chassis.WHEEL_PULSES_PER_REVOLUTION;
```
Comment (oldgalileo) [inline src/org/usfirst/frc4904/robot/RobotMap.java]: This is a constant and should thus be used as a constant. Furthermore, you should be putting this where it is used as a constant in all caps, and not in RobotMap.

File: src/org/usfirst/frc4904/robot/RobotMap.java [hunk 2]
```diff
@@ -57,55 +68,89 @@
 		
 		public static class PCM {}
 		
-		public static class Pneumatics {}
+		public static class Pneumatics {
```
Comment (oldgalileo) [inline src/org/usfirst/frc4904/robot/RobotMap.java:75]: There was intent behind removing these constants from RobotMap. We are trying to move away from having a jumble of constants in RobotMap which cause merge conflicts and make the code unorganized. The constants should be broken up and moved into the various classes in which they are used, as `public static final ALL_CAPS` fields. 

The other intention was to reduce clutter and slowly add back what we need. As we do not have a robot yet, there is not much that should be known yet. Some of the driver configurations could be retained, but if so they should be retained in a separate class. Potentially something along the lines of `DriverConfig` or something. I believe this is implemented per the new style in the terminator branch.

File: src/org/usfirst/frc4904/robot/humaninterface/drivers/HardMode.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+	
+	@Override
+	public double getX() {
+		return 0;
+	}
+	
+	@Override
+	public double getY() {
+		if (RobotMap.HumanInput.Driver.xbox.x.get()) {
+			return (RobotMap.HumanInput.Driver.xbox.lt.getX() + RobotMap.HumanInput.Driver.xbox.rt.getX() - 1) * RobotMap.Constant.HumanInput.Y_SPEED_SCALE;
```
Comment (oldgalileo) [inline src/org/usfirst/frc4904/robot/humaninterface/drivers/HardMode.java]: Why are you using the difference in the triggers for moving straight? As I understand it, this is a common control scheme for turning, but I don't quite understand it for lateral movement.

File: src/org/usfirst/frc4904/robot/humaninterface/drivers/HardMode.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+		} else {
+			return 0;
```
Comment (oldgalileo) [inline src/org/usfirst/frc4904/robot/humaninterface/drivers/HardMode.java]: Why?

File: src/org/usfirst/frc4904/robot/humaninterface/drivers/NathanGain.java [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+		super("NathanGain");
+	}
+	
+	protected double scaleGain(double input, double gain, double exp) {
+		return Math.pow(input, exp) * gain * Math.signum(input);
+	}
+	
+	@Override
+	public void bindCommands() {
+		Command normalDrive = new ChassisMove(RobotMap.Component.chassis, this);
```
Comment (oldgalileo) [inline src/org/usfirst/frc4904/robot/humaninterface/drivers/NathanGain.java]: Unused and thus unnecessary. Please remove.

**Review by oldgalileo — CHANGES_REQUESTED** [review]
See previous review.

File: src/org/usfirst/frc4904/robot/humaninterface/drivers/HardMode.java (hunk 3 above)
Comment (carturn) [reply src/org/usfirst/frc4904/robot/humaninterface/drivers/HardMode.java]: @HowardStark There was a driving mode implemented about a year ago that did use this control scheme, each trigger effectively controls one side of the robot.
However, **this control scheme was found to be very unsafe and not useful, and should be removed as soon as possible.**

**Comment by billwpierce** [comment]
@HowardStark please review

File: src/org/usfirst/frc4904/robot/RobotMap.java [hunk 6]
```diff
@@ -57,55 +69,79 @@
...
+			public static int solenoidUp = 0;
+			public static int solenoidDown = 1;
+		}
+	}
+	
+	public static class Constant {
```
Comment (oldgalileo) [inline src/org/usfirst/frc4904/robot/RobotMap.java]: This needs to me moved. See the other comment for reasons.

Comment (oldgalileo) [reply src/org/usfirst/frc4904/robot/RobotMap.java]: Suggested: Remove the Constants class completely, including all chassis variables. We do not use them yet.

File: src/org/usfirst/frc4904/robot/RobotMap.java (hunk 1 above)
Comment (oldgalileo) [reply src/org/usfirst/frc4904/robot/RobotMap.java]: Still not fixed.

File: src/org/usfirst/frc4904/robot/humaninterface/drivers/DefaultDriver.java [hunk 7]
```diff
@@ -4,7 +4,7 @@
 import org.usfirst.frc4904.standard.humaninput.Driver;
 
 public class DefaultDriver extends Driver {
-	public static final double CONTROLLER_MIN_THRESH = 0.1;
+	public static final double XBOX_MINIMUM_THRESHOLD = 0.1;
```
Comment (oldgalileo) [inline src/org/usfirst/frc4904/robot/humaninterface/drivers/DefaultDriver.java:6]: Good name change.

File: src/org/usfirst/frc4904/robot/RobotMap.java (hunk 1 above)
Comment (oldgalileo) [reply src/org/usfirst/frc4904/robot/RobotMap.java]: Suggested: Delete this. It isn't used.

File: src/org/usfirst/frc4904/robot/RobotMap.java [hunk 8]
```diff
@@ -57,55 +69,79 @@
...
+			public static double TURN_P = 0.02;
+			public static double TURN_I = 0.001;
+			public static double TURN_D = 0.3;
+			public static final double MAX_DEGREES_PER_SECOND = 120;
+			public static final double WHEEL_RADIUS_INCHES = 8;
+			public static final double WHEEL_PULSES_PER_REVOLUTION = 1024;
+		}
+	}
+	
+	public static class HumanInput {
```
Comment (oldgalileo) [inline src/org/usfirst/frc4904/robot/RobotMap.java]: Move HumanInput into Components.

File: src/org/usfirst/frc4904/robot/RobotMap.java [hunk 9]
```diff
@@ -57,55 +69,79 @@
...
+		public static class Driver {
+			public static CustomXbox xbox;
+		}
+		
+		public static class Operator {
+			public static CustomJoystick stick;
+		}
 	}
```
Comment (oldgalileo) [inline src/org/usfirst/frc4904/robot/RobotMap.java]: At this level, make a new static class called `Metrics`, and put in `WHEEL_PULSES_PER_REVOLUTION`

Comment (oldgalileo) [reply src/org/usfirst/frc4904/robot/RobotMap.java]: There should be only one

File: src/org/usfirst/frc4904/robot/RobotMap.java [hunk 10]
```diff
@@ -57,55 +69,79 @@
...
-		// Ball Dumper
-		Component.ballDumper = new Dump(new VictorSP(Port.PWM.vomitElevator), new VictorSP(Port.PWM.vomitOuttakeRoller));
-		// Flywheel
-		Motor leftFlywheelMotor = new Motor(new VictorSP(Port.PWM.flywheelLeftMotor));
-		leftFlywheelMotor.setInverted(true);
-		Motor rightFlywheelMotor = new Motor(new VictorSP(Port.PWM.flywheelRightMotor));
-		rightFlywheelMotor.setInverted(false);
-		Component.flywheel = new Flywheel(leftFlywheelMotor, rightFlywheelMotor);
+		double inchesPerPulse = 2 * Math.PI * Constant.Chassis.WHEEL_RADIUS_INCHES / Constant.Chassis.WHEEL_PULSES_PER_REVOLUTION;
+		Component.leftWheelEncoder.setDistancePerPulse(inchesPerPulse);
```
Comment (oldgalileo) [inline src/org/usfirst/frc4904/robot/RobotMap.java]: Use WHEEL_PULSES_PER_REVOLUTION for the value here.

**Review by oldgalileo — CHANGES_REQUESTED** [review]

File: src/org/usfirst/frc4904/robot/RobotMap.java (hunk 2 above)
Comment (billwpierce) [reply src/org/usfirst/frc4904/robot/RobotMap.java:75]: done mostly

File: src/org/usfirst/frc4904/robot/humaninterface/drivers/NathanGain.java [hunk 11]
```diff
@@ -0,0 +1,52 @@
+package org.usfirst.frc4904.robot.humaninterface.drivers;
+
+
+import org.usfirst.frc4904.robot.RobotMap;
+import org.usfirst.frc4904.standard.commands.chassis.ChassisShift;
+import org.usfirst.frc4904.standard.humaninput.Driver;
+import org.usfirst.frc4904.standard.subsystems.chassis.SolenoidShifters;
+
+public class NathanGain extends Driver {
+	public static final double SPEED_GAIN = 1;
```
Comment (oldgalileo) [inline src/org/usfirst/frc4904/robot/humaninterface/drivers/NathanGain.java:10]: Good placement.

File: src/org/usfirst/frc4904/robot/RobotMap.java [hunk 12]
```diff
@@ -28,84 +32,110 @@
...
+		}
 	}
-	
+
+	public static class Metrics {
+		public static final double WHEEL_PULSES_PER_REVOLUTION = 1024;
+	}
+
 	public static class Component {
+		public static class HumanInput {
```
Comment (oldgalileo) [inline src/org/usfirst/frc4904/robot/RobotMap.java]: HumanInput class shouldn't exist. The values it contains should be renamed and stuck in the root level of the Component class.

File: src/org/usfirst/frc4904/robot/RobotMap.java [hunk 13]
```diff
@@ -28,84 +32,110 @@
...
 		Component.operatorStick.setDeadzone(DefaultOperator.JOYSTICK_MIN_THRESH);
-		Component.driverXbox = new CustomXbox(Port.HumanInput.xboxController);
-		Component.driverXbox.setDeadZone(DefaultDriver.CONTROLLER_MIN_THRESH);
+		Component.HumanInput.Driver.xbox = new CustomXbox(Port.HumanInput.xboxController);
+		Component.HumanInput.Operator.stick = new CustomJoystick(Port.HumanInput.joystick);
+		Component.HumanInput.Driver.xbox.setDeadZone(DefaultDriver.XBOX_MINIMUM_THRESHOLD);
 		// Main Subsystems
 		Component.alignCamera = new AligningCamera(PIDSourceType.kRate);
-		Component.mainSubsystems = new Subsystem[] {Component.ballIntake, Component.ballDumper, Component.flywheel};
+		Component.mainSubsystems = new Subsystem[] {};// Component.ballIntake, Component.ballDumper, Component.flywheel};
```
Comment (oldgalileo) [inline src/org/usfirst/frc4904/robot/RobotMap.java]: Why is this empty?

**Review by oldgalileo — CHANGES_REQUESTED** [review]

**Review by oldgalileo — APPROVED** [review]
Good changes.

**Comment by billwpierce** [comment]
IT HAS BEEN TESTED --> Please merge.

**Pull request merged**